File: solutions/csharp/custom-set/1/CustomSet.cs
public class CustomSet
{
    private readonly HashSet<int> values;

    public CustomSet(params int[] values) => this.values = values.ToHashSet();

    public CustomSet Add(int value)
    {
        values.Add(value);
        return this;
    }

    public bool Empty() => values.Count == 0;

    public bool Contains(int value) => values.Contains(value);

    public bool Subset(CustomSet right) => values.IsSubsetOf(right.values);

    public bool Disjoint(CustomSet right) => !values.Overlaps(right.values);

    public CustomSet Intersection(CustomSet right) => new([.. values.Intersect(right.values)]);

    public CustomSet Difference(CustomSet right) => new([.. values.Except(right.values)]);

    public CustomSet Union(CustomSet right) => new([.. values.Union(right.values)]);

    public override bool Equals(object? obj) => obj is CustomSet other && values.SetEquals(other.values);

    public override int GetHashCode() => values.Aggregate(0, HashCode.Combine);
}

File: solutions/fsharp/custom-set/1/CustomSet.fs
module CustomSet

type Set<'a when 'a: equality> =
    private
    | Set of 'a list

let empty = Set []

let singleton value = Set [value]

let isEmpty (Set values) = List.isEmpty values

let size (Set values) = List.length values

let fromList values = Set (List.distinct values)

let toList (Set values) = values

let contains value (Set values) = List.contains value values

let insert value set =
    if contains value set then set else Set (value :: toList set)

let union left right =
    fromList (toList left @ toList right)

let intersection left right =
    toList left
    |> List.filter (fun value -> contains value right)
    |> fromList

let difference left right =
    toList left
    |> List.filter (fun value -> not (contains value right))
    |> fromList

let isSubsetOf left right =
    toList left |> List.forall (fun value -> contains value right)

let isDisjointFrom left right =
    toList left |> List.forall (fun value -> not (contains value right))

let isEqualTo left right =
    isSubsetOf left right && isSubsetOf right left
